pkgs.formats.toml: cryptic “Error writing TOML: unexpected primitive type …” when list mixes strings and attrsets (yj backend)

### Nixpkgs version

- Unstable (26.05)

### Describe the bug

`pkgs.formats.toml { }.generate` fails when a list contains both strings and attrsets (encoded as TOML inline tables). This shape is valid TOML (e.g. Helix `languages.toml` uses it), but the current generator backend errors during the build.

This is a regression: the same example succeeds on an older nixpkgs rev.

### Steps to reproduce

Create a directory with this `flake.nix`:

```nix
{
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/b12141ef619e0a9c1c84dc8c684040326f27cdcc";

  outputs = { self, nixpkgs }:
    let
      system = "aarch64-darwin";
      pkgs = import nixpkgs { inherit system; };
    in {
      packages.${system}.default =
        (pkgs.formats.toml { }).generate "repro.toml" {
          a = [
            "bash-language-server"
            { name = "typescript-language-server"; except-features = [ "diagnostics" ]; }
          ];
        };
    };
}
```

Run:

`nix build -L .`

Build fails with:

```
repro.toml> Error writing TOML: unexpected primitive type: *struct { A []interface {} "toml:\"except-features\""; B string "toml:\"name\"" }
error: builder for '/nix/store/...-repro.toml.drv' failed with exit code 1;
```

### Expected behaviour

Build should succeed and write a TOML file equivalent to:

```toml
a = ["bash-language-server", { name = "typescript-language-server", except-features = ["diagnostics"] }]
```

### Screenshots

_No response_

### Relevant log output

```console

```

### Additional context

https://github.com/NixOS/nixpkgs/pull/510585

### System metadata

 - system: `"aarch64-darwin"`
 - host os: `Darwin 25.4.0, macOS 26.4.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Lix, like Nix) 2.95.1
System type: aarch64-darwin
Additional system types: x86_64-darwin
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /Users/user/.config/nix/nix.conf:/Users/user/.nix-profile/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/d739a7g8397gk6c1k2vrcq6mym96hv0j-lix-2.95.1/share`
 - channels(root): `""`
 - nixpkgs: `/nix/store/qmqfjcv57sha8g976zqycdkvairb461j-source`

### Notify maintainers



---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

The `yj` backend switch introduced a similar error for me, except unfortunately i can't even get a stack origin from it.

```
❯ nix log /nix/store/ni8p5n1925ydd96mr3lqmwllp1mcgggm-helix-languages-config.drv
panic: eTable: unhandled reflect.Value Kind: string [recovered, repanicked]

goroutine 1 [running]:
github.com/BurntSushi/toml.(*Encoder).safeEncode.func1()
        github.com/BurntSushi/toml@v1.1.0/encode.go:131 +0x7c
panic({0x104e7a340?, 0x39374aa38d0?})
        runtime/panic.go:860 +0x12c
github.com/BurntSushi/toml.(*Encoder).eMapOrStruct(0x39374a31dd8, {0x39374aa4760, 0x2, 0x2}, {0x104e801a0?, 0x39374aa45d0?, 0x104f82800?}, 0x0)
        github.com/BurntSushi/toml@v1.1.0/encode.go:329 +0xec
github.com/BurntSushi/toml.(*Encoder).eArrayOfTables(0x39374a31dd8, {0x39374aa4760, 0x2, 0x2}, {0x104e78c60?, 0x39374ac2710?, 0x400a31101?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:304 +0x1d8
github.com/BurntSushi/toml.(*Encoder).encode(0x39374a31dd8, {0x39374aa4760, 0x2, 0x2}, {0x104e78c60?, 0x39374ac2710?, 0x104e78ae0?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:164 +0x2d8
github.com/BurntSushi/toml.(*Encoder).eStruct.func2({0x39374b42840?, 0x2, 0x39374ac0420?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:479 +0x2d0
github.com/BurntSushi/toml.(*Encoder).eStruct(0x39374a31dd8, {0x39374aa3070, 0x1, 0x1}, {0x39374ac0420?, 0x39374ac2700?, 0x39374a31501?}, 0x0)
        github.com/BurntSushi/toml@v1.1.0/encode.go:488 +0x19c
github.com/BurntSushi/toml.(*Encoder).eMapOrStruct(0x39374a31dd8, {0x39374aa3070, 0x1, 0x1}, {0x104e801a0?, 0x39374b42010?, 0x104f82800?}, 0x0)
        github.com/BurntSushi/toml@v1.1.0/encode.go:326 +0x80
github.com/BurntSushi/toml.(*Encoder).eArrayOfTables(0x39374a31dd8, {0x39374aa3070, 0x1, 0x1}, {0x104e78c60?, 0x39374aa4640?, 0x200a31601?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:304 +0x1d8
github.com/BurntSushi/toml.(*Encoder).encode(0x39374a31dd8, {0x39374aa3070, 0x1, 0x1}, {0x104e78c60?, 0x39374aa4640?, 0x104e78ae0?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:164 +0x2d8
github.com/BurntSushi/toml.(*Encoder).eStruct.func2({0x39374b426f0?, 0x2, 0x39374ac07e0?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:479 +0x2d0
github.com/BurntSushi/toml.(*Encoder).eStruct(0x39374a31dd8, {0x104ef8e80, 0x0, 0x0}, {0x39374ac07e0?, 0x39374aa4640?, 0x39374a31ae8?}, 0x0)
        github.com/BurntSushi/toml@v1.1.0/encode.go:488 +0x19c
github.com/BurntSushi/toml.(*Encoder).eMapOrStruct(0x39374a31dd8, {0x104ef8e80, 0x0, 0x0}, {0x39374ac07e0?, 0x39374aa4640?, 0x104e83ca0?}, 0x0)
        github.com/BurntSushi/toml@v1.1.0/encode.go:326 +0x80
github.com/BurntSushi/toml.(*Encoder).eTable(0x39374a31dd8, {0x104ef8e80, 0x0, 0x0}, {0x39374ac07e0?, 0x39374aa4640?, 0x0?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:318 +0x144
github.com/BurntSushi/toml.(*Encoder).encode(0x39374a31dd8, {0x104ef8e80, 0x0, 0x0}, {0x39374ac07e0?, 0x39374aa4640?, 0x0?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:184 +0x344
github.com/BurntSushi/toml.(*Encoder).safeEncode(0x39374ac27c0?, {0x104ef8e80?, 0xf0000000000001?, 0x39374ac27c0?}, {0x39374ac07e0?, 0x39374aa4640?, 0x0?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:134 +0x54
github.com/BurntSushi/toml.(*Encoder).Encode(0x39374a31dd8, {0x39374ac27c0?, 0x39374aa4640?})
        github.com/BurntSushi/toml@v1.1.0/encode.go:118 +0xa4
github.com/sclevine/yj/v5/convert.TOML.Encode({0x39374a31e38?, 0x14?}, {0x104eac558, 0x39374a82030?}, {0x104e8cc60, 0x39374a9e540})
        github.com/sclevine/yj/v5/convert/toml.go:34 +0x26c
main.Run({0x104eac538, 0x39374a82028}, {0x104eac558, 0x39374a82030}, {0x104eac558, 0x39374a82038}, {0x39374aa4140?, 0x104c20b98?, 0x104eb5be8?})
        github.com/sclevine/yj/v5/main.go:71 +0x280
main.main()
        github.com/sclevine/yj/v5/main.go:43 +0x5c
```

Built with [this](https://gitlab.com/dinckelman/nix/-/raw/main/modules/common/dev/editors/helix/lsp.mod.nix) nix object

--- a/pkgs/by-name/go/go-toml/package.nix
+++ b/pkgs/by-name/go/go-toml/package.nix
@@ -25,10 +25,6 @@ buildGoModule {
     "cmd/tomltestgen"
   ];
 
-  # allowGoReference adds the flag `-trimpath` which is also denoted by, go-toml's goreleaser config
-  #  <https://github.com/pelletier/go-toml/blob/a3d5a0bb530b5206c728eed9cb57323061922bcb/.goreleaser.yaml#L13>
-  allowGoReference = true;
-
   ldflags = [
     "-s"
     "-w"

--- a/pkgs/pkgs-lib/formats.nix
+++ b/pkgs/pkgs-lib/formats.nix
@@ -461,16 +461,18 @@ optionalAttrs allowAliases aliases
       generate =
         name: value:
         pkgs.callPackage (
-          { runCommand, yj }:
+          { runCommand, go-toml }:
           runCommand name
             {
-              nativeBuildInputs = [ yj ];
+              nativeBuildInputs = [ go-toml ];
               value = builtins.toJSON value;
               passAsFile = [ "value" ];
               preferLocalBuild = true;
             }
+            # -use-json-number: preserve JSON ints as TOML ints
+            # (Go's json.Decoder defaults to float64 for all numbers)
             ''
-              yj -jt < "$valuePath" > "$out"
+              jsontoml -use-json-number < "$valuePath" > "$out"
             ''
         ) { };
 

--- a/pkgs/pkgs-lib/tests/formats.nix
+++ b/pkgs/pkgs-lib/tests/formats.nix
@@ -695,16 +695,16 @@ runBuildTests {
       float = 3.141
       int = 10
       list = [1, 2]
-      str = "foo"
+      str = 'foo'
       true = true
 
       [attrs]
-      foo = "foo"
+      foo = 'foo'
 
       [level1]
       [level1.level2]
       [level1.level2.level3]
-      level4 = "deep"
+      level4 = 'deep'
     '';
   };
 

--- a/pkgs/pkgs-lib/tests/formats.nix
+++ b/pkgs/pkgs-lib/tests/formats.nix
@@ -708,6 +708,41 @@ runBuildTests {
     '';
   };
 
+  # Regression test for https://github.com/NixOS/nixpkgs/issues/511970
+  # yj crashes on arrays mixing scalars and attrsets (heterogeneous arrays),
+  # e.g. Helix language-server configs like ["bash-ls", {name = "ts-ls"; ...}].
+  # TOML 1.0 allows mixed-type arrays; the converter must emit them as
+  # inline arrays with inline tables.
+  tomlHeterogeneousArray = shouldPass {
+    format = formats.toml { };
+    input = {
+      language-server = [
+        "bash-language-server"
+        {
+          name = "typescript-language-server";
+          except-features = [ "diagnostics" ];
+        }
+      ];
+    };
+    expected = ''
+      language-server = ['bash-language-server', {except-features = ['diagnostics'], name = 'typescript-language-server'}]
+    '';
+  };
+
+  # Regression test for https://github.com/sclevine/yj/issues/52
+  # yj truncates keys at the first comma because it stores TOML keys in Go
+  # struct tags, where commas are option separators.
+  # e.g. "stack(x,n)" is emitted as "stack(x" — silently losing data.
+  tomlCommaInKey = shouldPass {
+    format = formats.toml { };
+    input = {
+      "stack(x,n)" = "foobar";
+    };
+    expected = ''
+      'stack(x,n)' = 'foobar'
+    '';
+  };
+
   cdnAtoms = shouldPass {
     format = formats.cdn { };
     input = {

